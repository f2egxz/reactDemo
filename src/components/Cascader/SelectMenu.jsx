// -*- mode： react-*-
// -*- coding: utf-8-*-

/**
 * Cascader.jsx
 *
 * 联级选择组件 的下拉框组件
 *
 * -TODO 下拉框列表
 *
 * @author fe.gxz
 */

import React              from 'react';
import { classnames }     from 'util/view-extra'
import MenuItem           from 'component/Cascader/SelectMenuitem.jsx'
import Styles             from 'style/Cascader.css'



class Menu extends React.Component {
  constructor(){
    super()
    this.state={
      stair:null,
      secondStage:null,
      threeStage:null,
      completed:false
    }
  }
  componentDidUpdate(){
    if(this.state.completed){
      const {onChange} = this.props
      const last = this.state.threeStage===null?'':this.state.threeStage+'/';
      const msg = this.state.stair+'/'+this.state.secondStage+'/'+last;
      onChange(msg)
    }
  }
  render(){
    const {options} =this.props
    const second = options.filter((item)=>(item.name===this.state.stair));
    const secondChildren = second[0]? second[0].sub:[]
    const three =  secondChildren.filter((item)=>(item.name===this.state.secondStage));
    const threeChildren = three[0]?three[0].sub:[]
    return (
      <div className = {classnames(Styles.menu)}>
        <MenuItem list = {options} choose={this.state.stair} onChange = {this.handleClickF.bind(this)}/>
        {this.state.stair!==null&&secondChildren!==null?  <MenuItem list={secondChildren} choose={this.state.secondStage} onChange = {this.handleClickS.bind(this)}/>:''}
        {this.state.secondStage!==null&&threeChildren!==null? <MenuItem list={threeChildren} choose={this.state.threeStage} onChange = {this.handleClickT.bind(this)}/>:''}
      </div>
    )
  }
  handleClickF(text,sub){
    this.setState({
      stair:text,
      secondStage:null,
      threeStage:null,
      completed:!sub
    })
  }
  handleClickS(text,sub){
    const {onChange} =this.props
    this.setState({
      secondStage:text,
      threeStage:null,
      completed:!sub
    })
  }
  handleClickT(text,sub){
    const {onChange} =this.props
    this.setState({
      threeStage:text,
      completed:!sub
    })
  }
}


export default Menu
