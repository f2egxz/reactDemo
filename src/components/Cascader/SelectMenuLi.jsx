// -*- coding: utf-8-*-

/**
 * Cascader.jsx
 *
 * 联级选择组件 的下拉框组件 的选项
 *
 * -TODO 下拉框选项
 *
 * @author fe.gxz
 */

import React              from 'react';
import { classnames }     from 'util/view-extra'
import Styles             from 'style/Cascader.css'

class SlelectMenuLi extends React.Component {
  constructor(){
    super()
  }
  render(){
    const { text,sub,choose } = this.props
    return (
      <li className={classnames(Styles.defaultLi,sub?Styles.rightArrow:'',choose===text?Styles.choose:'')}
        onClick={this.handleClick.bind(this)}>{text}</li>
    )
  }
  handleClick(){
    const { text,onChange,sub } = this.props
    onChange(text,sub)
  }
}

export default SlelectMenuLi
