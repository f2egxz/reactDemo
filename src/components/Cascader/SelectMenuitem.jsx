// -*- coding: utf-8-*-

/**
 * Cascader.jsx
 *
 * 联级选择组件 的下拉框组件 的菜单
 *
 * -TODO 下拉框子菜单
 *
 * @author fe.gxz
 */

import React              from 'react';
import { classnames }     from 'util/view-extra'
import Styles             from 'style/Cascader.css'
import Mli                from 'component/Cascader/SelectMenuLi.jsx'

function MenuItem ({
  list,onChange,choose
}){
  const lis = list.map((item)=>(
    <Mli sub = {Array.isArray(item.sub)&&item.sub.length>0? true:false}
         key = {item.name}
         text = {item.name}
         choose = {choose}
         onChange = {onChange}
    />
  ))
  return (
    <ul className={classnames(Styles.defaultUl)}>
      {lis}
    </ul>
  )
}


export default MenuItem
