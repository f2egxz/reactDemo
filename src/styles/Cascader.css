/*-*-mode: react-*-*/
/*-*-coding: utf-8-*-*/


/**
 * Cascader.css
 *
 * 联级选择器组件
 *
 * @auther fe.gxz
 */

@value  choose :   #f7f7f7;
@value  hover  :   #ecf6fd;

.defaultUl {
  display: inline-block;
  vertical-align: top;
  min-width: 111px;
  height: 180px;
  list-style: none;
  margin: 0;
  padding: 0;
  border-right: 1px solid #e9e9e9;
  overflow: auto;
  border-radius: 4px 0 0 4px;
}

.defaultLi {
  position: relative;
  padding: 7px 26px 7px 16px;
  cursor: pointer;
  white-space: nowrap;
   &:hover {
      background:  #ecf6fd;
   }
}

.rightArrow:after {
   content: '>';
   position: absolute;
   top:8px;
   right: 18px;
 }

.menu {
  font-size: 12px;
  background: #fff;
  position: absolute;
  z-index: 1050;
  border-radius: 4px;
  box-shadow: 0 1px 6px rgba(0, 0, 0, 0.2);
  white-space: nowrap;
}

.choose {
  background-color: #f7f7f7;
}

.inputBox{
  display: inline-block;
  position: relative;
  cursor: pointer;
  font-size: 12px;
}

.text {
  position: absolute;
  left: 0;
  line-height: 20px;
  top: 50%;
  margin-top: -10px;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  padding: 0 12px 0 8px;
  width: 100%;
  height: 100%;
  &::before {
     content: '▽';
     position: absolute;
     top:1px;
     right: 26px;
     font-size: 20px;
  }
}
.close{
  position: absolute;
  top:3px;
  right: 6px;
  height: 20px;
  width: 20px;
  background-color: #ccc;
  border-radius: 100%;
  text-align: center;
}



