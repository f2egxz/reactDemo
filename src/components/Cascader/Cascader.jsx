// -*- mode： react-*-
// -*- coding: utf-8-*-

/**
 * Cascader.jsx
 *
 * 联级选择组件
 *
 * -TODO 联级选择
 *
 * @author fe.gxz
 */

import React              from 'react';
import { classnames }     from 'util/view-extra'
import Styles             from 'style/Cascader.css'
import SelectMenu         from 'component/Cascader/SelectMenu.jsx'

/**
 * Cascader
 *
 * 联级选择组件
 *
 * @param {Array} options
 * @param {function} onChange
 *
 */
class Cascader extends React.Component {
  constructor(){
    super()
    this.state = {
      show:false,
      choose:'',
      close:false,
      placeholder:'Please Select',
    }
  }
  render(){
    const {options} = this.props;
    return (
      <div>
        <span className={classnames(Styles.inputBox)}
              onMouseEnter={()=>this.setState({close:true})}
              onMouseLeave={()=>this.setState({close:false})}
        >
          <input type="text" placeholder={this.state.placeholder}/>
          <span className={classnames(Styles.text)}
                onClick={this.handleClickShow.bind(this)}
          >{this.state.choose}</span>
          { this.state.choose !== '' && this.state.close?
            <i className = {Styles.close} onClick={this.handleClose.bind(this)}>x</i>
            : ''
          }
        </span>
        {this.state.show ? <SelectMenu options={options} sub = {true} onChange = {this.handleClick.bind(this)}/>:''}
        
      </div>
    )
  }
  handleClick(text){
    const {onChange} =this.props
    this.setState({
      choose:text,
      placeholder:'',
      show:false
    })
    onChange(text)
  }
  handleClickShow(){
    this.setState({
      show:!this.state.show
    })
  }
  handleClose(){
    const {onChange} =this.props
    this.setState({
      show:false,
      choose:'',
      close:false,
      placeholder:'Please Select',
    })
    onChange('')
  }
}

export default Cascader





