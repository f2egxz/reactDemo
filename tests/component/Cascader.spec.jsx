import React from 'react'
import Cascader from 'component/Cascader/Cascader.jsx'
import render from 'react-test-renderer'

test('Cascader',()=>{
  const handleClick = jest.fn()
  const data = [ {
    "name": "北京",
    "sub": [
      {
        "name": "北京",
        "sub": [
          {
            "name": "东城区"
          },
          {
            "name": "西城区"
          },
          {
            "name": "崇文区"
          },
          {
            "name": "宣武区"
          },
          {
            "name": "朝阳区"
          },
          {
            "name": "丰台区"
          },
          {
            "name": "石景山区"
          },
          {
            "name": "海淀区"
          },
          {
            "name": "门头沟区"
          },
          {
            "name": "房山区"
          },
          {
            "name": "通州区"
          },
          {
            "name": "顺义区"
          },
          {
            "name": "昌平区"
          },
          {
            "name": "大兴区"
          },
          {
            "name": "怀柔区"
          },
          {
            "name": "平谷区"
          },
          {
            "name": "密云县"
          },
          {
            "name": "延庆县"
          }
        ]
      }
    ]
  }]
  const component = render.create(
    <Cascader options={data} onChange = {handleClick}/>
  )
  let CascaderString = component.toJSON;
  expect(CascaderString).toMatchSnapshot();
})
