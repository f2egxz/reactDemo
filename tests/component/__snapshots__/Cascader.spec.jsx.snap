exports[`test Cascader 1`] = `[Function]`;
